src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,375 @@
+      failedResult.caughtException = true;
+      result = kj::mv(failedResult);
+    });
+    return kj::mv(KJ_REQUIRE_NONNULL(result));
   }
+
+  // Invokes the given (synchronous) function, and serializes the result for RPC.
+  SerializedResult prepareSyncResult(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    kj::Maybe<v8::Local<v8::Value>> result;
+    bool caughtException = false;
+    js.tryCatch([&] {
+      result = invokeFn(js, fn, serializedArgs);
+    }, [&](jsg::Value exception) {
+      // We threw while executing JS.
+      caughtException = true;
+      result = exception.getHandle(js);
+    });
+
+    auto jsVal = jsg::JsValue(KJ_REQUIRE_NONNULL(result));
+    auto serializedResult = serialize(js, kj::mv(jsVal));
+
+    // `caughtException` might be true if the JS method threw an exception, so we don't want to
+    // accidentally set it to false just because serialization succeeded.
+    if (caughtException) {
+      serializedResult.caughtException = caughtException;
+    }
+    return kj::mv(serializedResult);
+  }
+
+  // Invokes the given (asynchronous) function, and serializes the result for RPC.
+  kj::Promise<SerializedResult> prepareAsyncResult(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    // We can't co_await a jsg::Promise, so we will use a promise-fulfiller pair instead.
+    auto [completionPromise, fulfiller] = kj::newPromiseAndFulfiller<SerializedResult>();
+    auto result = invokeFn(js, fn, serializedArgs);
+    auto jsPromise = js.toPromise(result.As<v8::Promise>())
+        .then(js, [this, &fulfiller=*fulfiller] (jsg::Lock& lock, jsg::Value value) mutable {
+          auto jsVal = jsg::JsValue(value.getHandle(lock));
+          fulfiller.fulfill(serialize(lock, kj::mv(jsVal)));
+        }).catch_(js, [this, &fulfiller=*fulfiller](jsg::Lock& lock, jsg::Value exception) mutable {
+          auto jsVal = jsg::JsValue(exception.getHandle(lock));
+          auto result = serialize(lock, kj::mv(jsVal));
+          // We are in the error branch, so regardless of if serialization succeeded we've
+          // caught an exception.
+          result.caughtException = true;
+
+          fulfiller.fulfill(kj::mv(result));
+        });
+
+    // Wait on the kj::Promise, which will resolve once the JS promise has resolved.
+    co_return co_await kj::mv(completionPromise);
+  }
+
+  // We use the callFulfiller to let the custom event know we've finished executing the method,
+  // but it also signals whether the call completed succesfully or if it failed (exception thrown).
+  kj::Own<kj::PromiseFulfiller<bool>> callFulfiller;
+  IoContext& ctx;
+  kj::Maybe<kj::StringPtr> entrypointName;
+};
+
+GetJsRpcTargetEvent::GetJsRpcTargetEvent()
+    : ExtendableEvent("getJsRpcTarget") {};
+
+kj::Promise<WorkerInterface::CustomEvent::Result> GetJsRpcTargetCustomEventImpl::run(
```

Comment:
At one point we discussed having a separate compat flag to enable RPC from the server side, so that existing apps in the wild don't suddenly find all the methods on their DOs have become remotely callable. Did we decide we wanted to do that or did we decide it doesn't really matter since the clients are necessarily on the same account? Personally I still think the compat flag would be a good idea.

Reply:
Note that this would _not_ be the `experimental` flag, this would be a new flag we expect apps actually have to enable to use RPC (but would eventually be default-on).

Reply:
Ahhh I see, you're saying another script that does not implement the DO but has a binding for it can have `experimental` on (so new stub is enabled), but the script that implements the DO does not, so it wouldn't expect an inbound JS RPC call but may receive one.

Having a flag on the server-side makes sense to me. I'll add that ASAP.

Reply:
https://github.com/cloudflare/workerd/compare/e51ab6d48ac02777186f0ef14219ba0f1407998f..e81607ef0fc9e144e1210efbe4c78d3d27cdfff3 introduces the `js_rpc` flag. 

I'm checking it in `call()` since we're in the isolate by that point so it's easy to get the compat flags. Also added another test to confirm that we fail the RPC call (on the server side) with a message if the DO doesn't have the `js_rpc` flag.

Reply:
Ok so I changed a couple things and got the `windows-debug` build to pass, going to slowly remove changes until the build breaks again. Unfortunately I don't have a windows machine so I have to run it in CI :slightly_frowning_face: